tests/always/test_connection.py

```diff
@@ -737,3 +737,6 @@ def test_extra_warnings_non_json(self):
     def test_extra_warnings_non_dict_json(self):
         with pytest.warns(DeprecationWarning, match='not parse as a dictionary'):
             Connection(conn_id='test_extra', conn_type='none', extra='"hi"')
+
+    def test_get_uri_no_conn_type(self):
+        assert Connection().get_uri() == 'none://'
```

Comment:
What's the usecase for this? Cos this doesn't seem like a useful URL, not to mention the column has this defined as `nullable=False`!

Reply:
this is not a useful URI of course...

i could have written the test as `Connection(host='blah', login='blah').get_uri() == 'none://blah@blah'` and the objection would not apply.  but, these other attrs have nothing to do with the test.  we're merely verifying that the URI generation doesn't fail and `none` is the scheme.

the point here is, if you just want to generate a URI for an airflow connection, let's make it easy.  airflow doesn't care whether your uri is `none://blah@blah` or `abc://blah@blah` -- it will still work when you set that conn in env vars or secrets backend.

Reply:
What I mean is: how can you hever have a null/none conn_type? That _should_ be an error somewhere as it is a required field.

Reply:
Yeah it's non-nullable in the db, but as for the python object it's not required.  

We could make it required but it's been this way a long time.  There are some hooks that accept `Connection` as an arg so if user is building the connection directly and passing it to hook and not specifying conn type it could break for them.  Though maybe [xkcd 1172](https://www.explainxkcd.com/wiki/index.php/1172:_Workflow) applies.    

Here's an example from our tests where we use a conn object with no conn type: https://github.com/apache/airflow/blob/d67ac5932dabbf06ae733fc57b48491a8029b8c2/tests/always/test_connection.py#L82

When you omit scheme from URI, it gets converted to empty string:
```
from airflow.models.connection import Connection

Connection(uri='//login:pass@host')
```

So we could always make it not quite required but give it a default of empty string. 

Let me know your thoughts.

Reply:
Add one more example to cover the json case...

When storing your conn as json, omitting conn type works (because it works for the python object and we're just forwarding the kwargs)

```
import os
os.environ['AIRFLOW_CONN_MY_CONN']= '{"login": "login", "password": "pass", "host": "host"}'
from airflow.hooks.base import BaseHook
BaseHook.get_connection('my_conn').conn_type
```

In this case the conn_type is None ,  just like the `Connection(extra='testextra')` example.

Reply:
My two cents:

of course Airflow doesn't care whether the uri is `none://blah@blah` or `abc://blah@blah`. However, what is the advantage or necessity of having `none://...`?

By making it compulsory to have the conn type, at least we are "forcing" the user to be careful with the values/names, which is good. And at the end of day, the connection is a URI, and it should follow the convention (RFC3986)

Reply:
`//login:pass@host` is a fully valid URI -- it's called a scheme-local URI. (One example: https://stackoverflow.com/questions/4387289/how-widely-supported-are-scheme-relative-uris-in-http-301-redirects)

So I'd vote for either making conn_type required (which is is when you do things "normally" via the UI) at the object level, or required when calling get_uri, or just allowing this edge case that type-less connections get scheme relative URIs out. 

-1 to `none://` though.

Reply:
> of course Airflow doesn't care whether the uri is none://blah@blah or abc://blah@blah. However, what is the advantage or necessity of having none://...

Well I'm not partial to `none://`... i think `//` is actually better for consistency.  that's just what came to mind as a placeholder for `None`.  The only use of `get_uri` is user convenience when producing a URI for use in env vars or secrets backend.  Not having to specify the conn type, when for your use case it doesn't matter (as indicated by the fact that you did not supply it, is more convenient than having to make up an arbitrary string that has no meaning and doesn't do anything for you.

> By making it compulsory to have the conn type, at least we are "forcing" the user to be careful with the values/names, which is good. 

Yes but it could also be that we're forcing them to jump through a hoop for no reason.  If they need a conn type, they need a conn type; if they don't, then they don't.

> And at the end of day, the connection is a URI, and it should follow the convention (RFC3986)

URI is one such representation but we also support yaml, json, storage in DB, and of course native python objects ;)

> or just allowing this edge case that type-less connections get scheme relative URIs out.

I'll run with this ;)